src/hooks/useClickAway/useClickAway.tsx

```diff
@@ -0,0 +1,48 @@
+'use client';
+
+import { useEffect, useRef } from 'react';
+
+import { CallbackType } from './useClickAway.type';
+
+const events = ['mousedown', 'touchstart'];
+
+const useClickAway = (callback: CallbackType) => {
+  const ref = useRef(null);
+  const saveCallback = useRef(callback);
+
+  useEffect(() => {
+    saveCallback.current = callback;
+  }, [callback]);
+
+  useEffect(() => {
+    const element = ref.current as unknown as HTMLElement;
+
+    if (!element) {
+      return;
+    }
+
+    const eventHandler = (e: Event) => {
+      const { target } = e;
+
+      if (element.contains(target as Node)) {
+        return;
+      }
+
+      saveCallback.current(e);
+    };
+
+    events.forEach((event) => {
+      document.addEventListener(event, eventHandler);
+    });
+
+    return () => {
+      events.forEach((eventName) => {
+        document.removeEventListener(eventName, eventHandler);
+      });
+    };
+  }, []);
+
+  return ref;
+};
```

Comment:
[ P2 ] `as` ~~~~~~~~~~~~?????

```suggestion
const useClickAway = <T,>(callback: CallbackType) => {
  const ref = useRef<T | null>(null);
  const saveCallback = useRef(callback);

  useEffect(() => {
    saveCallback.current = callback;
  }, [callback]);

  useEffect(() => {
    const element = ref.current;

    if (!element || !(element instanceof HTMLElement)) {
      return;
    }

    const eventHandler = (e: MouseEvent | TouchEvent) => {
      const { target } = e;

      if (!target || !(target instanceof HTMLElement)) {
        return;
      }

      if (element.contains(target)) {
        return;
      }

      saveCallback.current(e);
    };

    events.forEach((event) => {
      document.addEventListener(event, eventHandler);
    });

    return () => {
      events.forEach((eventName) => {
        document.removeEventListener(eventName, eventHandler);
      });
    };
  }, []);

  return ref;
};
```

Reply:
타입 관련 많이 배워갑니다..